numpy/core/src/multiarray/ctors.c

```diff
@@ -2181,42 +2180,9 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
 
     /* Raise an error if the casting rule isn't followed */
     if (!PyArray_CanCastArrayTo(arr, newtype, casting)) {
-        PyObject *errmsg;
-        PyArray_Descr *arr_descr = NULL;
-        PyObject *arr_descr_repr = NULL;
-        PyObject *newtype_repr = NULL;
-
         PyErr_Clear();
-        errmsg = PyUString_FromString("Cannot cast array data from ");
-        arr_descr = PyArray_DESCR(arr);
-        if (arr_descr == NULL) {
-            Py_DECREF(newtype);
-            Py_DECREF(errmsg);
-            return NULL;
-        }
-        arr_descr_repr = PyObject_Repr((PyObject *)arr_descr);
-        if (arr_descr_repr == NULL) {
-            Py_DECREF(newtype);
-            Py_DECREF(errmsg);
-            return NULL;
-        }
-        PyUString_ConcatAndDel(&errmsg, arr_descr_repr);
-        PyUString_ConcatAndDel(&errmsg,
-                PyUString_FromString(" to "));
-        newtype_repr = PyObject_Repr((PyObject *)newtype);
-        if (newtype_repr == NULL) {
-            Py_DECREF(newtype);
-            Py_DECREF(errmsg);
-            return NULL;
-        }
-        PyUString_ConcatAndDel(&errmsg, newtype_repr);
-        PyUString_ConcatAndDel(&errmsg,
-                PyUString_FromFormat(" according to the rule %s",
-                        npy_casting_to_string(casting)));
-        PyErr_SetObject(PyExc_TypeError, errmsg);
-        Py_DECREF(errmsg);
-
-        Py_DECREF(newtype);
+        npy_set_invalid_cast_error(
+                PyArray_DESCR(arr), newtype, casting, PyArray_NDIM(arr) == 0);
```

Comment:
Doesn't this now describe a 0d array as a scalar?

Reply:
Yes, but maybe no... that is what `PyArray_CanCastArrayTo` does. Its a bit annoying, it can be incorrect from the user-input perspective, but more correct due to value based casting of 0-D arrays. I.e. the 0-D array uses "scalar casting" rules right now, but the original input typically could have been either a scalar or an array.

Reply:
What's the code path that hits this error?

Reply:
Pretty much none... The only paths that should hit this, are arguably nonsense:
```
np.take([2], [1], out=np.array([1], dtype=np.uint64))
```
but even that example cannot get a "scalar" case.
```
from numpy.core import _rational_tests as rt                                                                   
np.array(np.array(rt.rational(2)), dtype="M8")  
```
should hit it IMO, but we currently consider all unsafe casts as possible, and in this case error when we notice that there is no way defined to do the actual casts. That is: In almost all cases we use force-casting, so it just doesn't matter.

Reply:
Presumably `np.array(rt.rational(2), dtype="M8")` is a case when the scalar message _is_ justified?

Reply:
Mind adding one of these as a test?

Reply:
I can add the test, but it is basically adding it for the future (or excersizing a different error). Because "unsafe" casts are currently always considered possible, it errors later with:
```
ValueError: No cast function available
```

Reply:
I added a test for a value error, I am not sure its useful here to be honest. But likely enough we never had a test for that in any case...

Reply:
Coverage is always useful!